### PR TITLE
Use binary search for searching extent

### PR DESCRIPTION
**Pull request by baekrang256**

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Run `clang-format -i` for consistent style.

File: extent.c [hunk 1]
```diff
@@ -6,19 +6,58 @@
 /* Search for the extent containing the target block.
  * Returns the first unused file index if not found.
  * Returns -1 if the target block is out of range.
- * TODO: Implement binary search for efficiency.
+ * Implemented binary search for efficiency
```
Comment (jserv) [inline extent.c]: Rephrase as following:
> Binary search is used for efficiency.

File: extent.c [hunk 2]
```diff
@@ -6,19 +6,58 @@
...
+    start = 0;
+    end = boundary - 1;
+    while (start < end) {
+        uint32_t mid = start + (end - start) / 2;
+        uint32_t block = index->extents[mid].ee_block;
+        uint32_t len = index->extents[mid].ee_len;
+        if (iblock >= block && iblock < block + len) {
+            /* found before search finished */
+            return mid;
+        } else if (iblock < block) {
```
Comment (jserv) [inline extent.c]: Remove `else`. That is,
```c
        if (iblock >= block && iblock < block + len) {
            /* found before search finished */
            return mid;
        }
        if (iblock < block)
            end = mid;
        else
            start = mid + 1;
```

File: extent.c [hunk 3]
```diff
@@ -6,19 +6,58 @@
...
+        } else {
+            start = mid + 1;
+        }
+    }
+
+    /* check that end directs to valid block. */
+    end_block = index->extents[end].ee_block;
+    end_len = index->extents[end].ee_len;
+
+    if (iblock >= end_block && iblock < end_len) {
```
Comment (jserv) [inline extent.c:68]: Ditto. Rewrite as following:
```c
    if (iblock >= end_block && iblock < end_len)
        return end;
    return boundary;
```

**Review by jserv — CHANGES_REQUESTED** [review]
Use `git rebase -i` to squash the commits and read the article [How to Write a Git Commit Message](https://cbea.ms/git-commit/) to refine the commit message.

**Comment by jserv** [comment]
@baekrang256, can you figure out the reason why GitHub Actions can not complete the tests?

**Comment by RoyWFHuang** [comment]
These are some bugs in this function. I dump the kernel message and it shows:
`[  333.089632] BUG: unable to handle page fault for address: ffff94e6dcb34ff8`
`[  333.099146] RIP: 0010:simplefs_ext_search.cold+0x46/0xa5 [simplefs]`

By the information 0x46, we can find 
`  46:	41 8b 78 04          	mov    0x4(%r8),%edi`
`        uint32_t len = index->extents[mid].ee_len;`
it seems that the "mid" is out of range, so you may need to check when the "end" is 0 and  assign to "boundary", which will make "end" reach to unsinged max 

Another thing, you should also consider when the "end" is MAX_EXTENTS and ee_start != 0, this will also cause the same error.

**Comment by baekrang256** [comment]
There were indeed some corner cases mentioned by @RoyWFHuang that was not handled. Thanks for the review.

I've fixed the code and tried the CI test in my local environment, checked the output comes out as expected according to other CI test outputs from other pr. I think you can try out the github action workflow again. 

One additional question : should I rebase the branch again before the merge?

**Comment by jserv** [comment]
> One additional question : should I rebase the branch again before the merge?

Yes, please go ahead.

**Comment by baekrang256** [comment]
Finished rebasing.

**Comment by jserv** [comment]
Thank @baekrang256 for contributing!
